### PR TITLE
BuildUpToDateCheck documentation and refactor

### PR DESCRIPTION
**Pull request by drewnoakes**
Adds `docs/up-to-date-check.md` that describes the process used by `BuildUpToDateCheck`.

Also some minor refactoring and code documentation.

### CODE REVIEW

File: docs/up-to-date-check.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# Up-to-date Check
```
Comment (Pilchie) [inline docs/up-to-date-check.md]: The normal convention here is that only docs that we expect to point end-users to go in this directory, and docs about how to work in the repo, the implementation, etc go in the `docs/repo` directory.  This seems like more of the latter.

File: docs/up-to-date-check.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+- [`BuildUpToDateCheck.cs`](https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs)
+- [`BuildUpToDateCheckTests.cs`](https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/UpToDate/BuildUpToDateCheckTests.cs)
+
+## Debugging
+
+The `BuildUpToDateCheck` class uses a `BuildUpToDateCheckLogger` to log verbose and info level information that is very useful during debugging.
+
+The log level is obtained via `IProjectSystemOptions.GetFastUpToDateLoggingLevelAsync`.
+
+TODO document how to view this log
```
Comment (Pilchie) [inline docs/up-to-date-check.md]: https://github.com/dotnet/project-system/blob/master/docs/repo/debugging/logging.md documents this already.

File: docs/up-to-date-check.md (hunk 1 above)
Comment (drewnoakes) [reply docs/up-to-date-check.md]: Makes sense. I'll move it in a subsequent PR as #4023 follows on from this PR and modifies the file.

File: docs/up-to-date-check.md (hunk 2 above)
Comment (drewnoakes) [reply docs/up-to-date-check.md]: I don't see any logging related to the up to date check in the "Project" output panel. I'll have a look at CPS to see what it does with the `TextWriter` it passes on the interface.

Comment (drewnoakes) [reply docs/up-to-date-check.md]: Am looking into this and will update once resolved.

**Pull request merged**